pkg/util/mount/nsenter_mount.go

```diff
@@ -0,0 +1,162 @@
+// 4.  Kubelet process has CAP_SYS_ADMIN -- at the present, this effectively
+//     means that the kubelet is running in a privileged container.
+type NsenterMounter struct{}
+
+// NsenterMounter implements mount.Interface
+var _ = Interface(&NsenterMounter{})
+
+const (
+	hostRootFsPath     = "/rootfs"
+	hostProcMountsPath = "/rootfs/proc/mounts"
+
+	HACK_nsenterPath = "/nsenter"
+)
+
+// Mount runs mount(8) in the host's root mount namespace.
+func (*NsenterMounter) Mount(source string, target string, fstype string, flags uintptr, data string) error {
+	glog.Infof("nsenter Mounting %s %s %s %d %s", source, target, fstype, flags, data)
+
+	args := makeArgs(source, target, fstype, flags, data)
+
+	glog.Infof("Mount command: %v %v", HACK_nsenterPath, args)
+	exec := exec.New()
+	outputBytes, err := exec.Command(HACK_nsenterPath, args...).CombinedOutput()
+	if len(outputBytes) != 0 {
+		glog.Infof("Output from mount command: %v", string(outputBytes))
+	}
+
+	return err
+}
+
+// TODO: remove once #6400 is in.
+func makeArgs(source, target, fstype string, flags uintptr, data string) []string {
+	args := []string{
+		"--mount=/rootfs/proc/1/ns/mnt",
+		"/usr/bin/mount",
+		"-t",
+		fstype,
+	}
+
+	// HACK: once 6400 is in, we can take this out
+	if flags&syscall.MS_BIND == syscall.MS_BIND {
+		data += ",bind"
+	}
+	if flags&syscall.MS_RDONLY == syscall.MS_RDONLY {
+		data += ",ro"
+	}
+	if data != "" {
+		args = append(args, "-o", data)
+	}
+
+	return append(args, source, target)
+}
+
+// Unmount runs umount(8) in the host's root mount namespace.
+func (*NsenterMounter) Unmount(target string, flags int) error {
+	args := []string{
+		"--mount=/rootfs/proc/1/ns/mnt",
+		"/usr/bin/umount",
+		target,
+	}
+
+	glog.Infof("Unmount command: %v %v", HACK_nsenterPath, args)
+	exec := exec.New()
+	outputBytes, err := exec.Command(HACK_nsenterPath, args...).CombinedOutput()
+	if len(outputBytes) != 0 {
+		glog.Infof("Output from mount command: %v", string(outputBytes))
+	}
+
+	return err
+}
+
+// List returns a list of all mounted filesystems in the host's mount namespace.
+func (*NsenterMounter) List() ([]MountPoint, error) {
+	hash1, err := readProcMounts(hostProcMountsPath, nil)
+	if err != nil {
+		return nil, err
+	}
+
+	for i := 0; i < maxListTries; i++ {
+		mps := []MountPoint{}
+		hash2, err := readProcMounts(hostProcMountsPath, &mps)
+		if err != nil {
+			return nil, err
+		}
+		if hash1 == hash2 {
+			// Success
+			return mps, nil
+		}
+		hash1 = hash2
+	}
+	return nil, fmt.Errorf("failed to get a consistent snapshot of %v after %d tries", hostProcMountsPath, maxListTries)
+}
+
+// IsMountPoint determines whether a path is a mountpoint by calling findmnt
+// in the host's root mount namespace.
+func (*NsenterMounter) IsMountPoint(file string) (bool, error) {
+	// The above is fast, but incorrect for bind mounts, so use findmnt
+	// We could look through /proc/self/mounts, but what a pain.
```

Comment:
@eparis heh, that's on my list to take out.  I'm going to rebase and make a big pass through tonight.